aten/src/TH/generic/THTensorRandom.cpp

```diff
@@ -477,32 +481,81 @@ void THTensor_(getRNGState)(at::Generator *_generator, THTensor *self)
-  static const size_t size = sizeof(at::CPUGenerator);
-  at::CPUGenerator* rng_state;
+  static const size_t size = sizeof(THGeneratorStateNew);
   THTensor_(resize1d)(self, size);
   THArgCheck(THTensor_(nElement)(self) == size, 1, "RNG state is wrong size");
   THArgCheck(THTensor_(isContiguous)(self), 1, "RNG state needs to be contiguous");
-  rng_state = (at::CPUGenerator*)self->data<scalar_t>();
+
+  // cast byte tensor to POD type
+  THGeneratorStateNew* rng_state = (THGeneratorStateNew*)self->data<scalar_t>();
+
+  // accumulate generator data to be copied into byte tensor
+  auto state = c10::guts::make_unique<THGeneratorStateNew>();
   auto cast_generator = static_cast<at::CPUGenerator*>(_generator);
-  memcpy(rng_state, cast_generator, size);
+  state->engine = cast_generator->engine();
+  state->next_float_normal_sample = cast_generator->next_float_normal_sample();
+  state->next_double_normal_sample = cast_generator->next_double_normal_sample();
+
+  memcpy(rng_state, state.get(), size);
 }
 
 void THTensor_(setRNGState)(at::Generator *_generator, THTensor *self)
 {
   // See Note [Thread-safety and Generators]
   std::lock_guard<std::mutex> lock(_generator->mutex_);
-  static const size_t size = sizeof(at::CPUGenerator);
-  at::CPUGenerator* rng_state;
-  THArgCheck(THTensor_(nElement)(self) == size, 1, "RNG state is wrong size");
+  auto cast_generator = static_cast<at::CPUGenerator*>(_generator);
   THArgCheck(THTensor_(isContiguous)(self), 1, "RNG state needs to be contiguous");
-  rng_state = (at::CPUGenerator*)self->data<scalar_t>();
-  int64_t is_valid = 0;
-  if (_generator->device().type() == rng_state->device().type()) {
-    is_valid = 1;
+  static const size_t size_legacy = sizeof(THGeneratorState);
+  static const size_t size_current = sizeof(THGeneratorStateNew);
+  at::mt19937 engine;
+  auto float_normal_sample = c10::optional<float>();
+  auto double_normal_sample = c10::optional<double>();
+
+  // Construct the state of at::CPUGenerator based on input byte tensor size.
+  if (THTensor_(nElement)(self) == size_legacy) {
+    THGeneratorState* legacy_state = (THGeneratorState*)self->data<scalar_t>();
+    // Note that in legacy THGeneratorState, we didn't have float version
+    // of normal sample and hence we leave the c10::optional<float> as is
+    
+    // Update next_double_normal_sample.
+    // Note that legacy THGeneratorState stores two uniform values (normal_x, normal_y)
+    // and a rho value (normal_rho). These three values were redundant and in the new
+    // DistributionsHelper.h, we store the actual extra normal sample, rather than three
+    // intermediate values.
+    if (legacy_state->normal_is_valid) {
+      auto r = legacy_state->normal_rho;
+      auto theta = 2.0 * M_PI * legacy_state->normal_x;
+      // we return the sin version of the normal sample when in caching mode
+      double_normal_sample = c10::optional<double>(r * ::sin(theta));
```

Comment:
Does this actually give you identical behavior to the old sampling behavior? I'm not convinced, because you only set `double_normal_sample` here, and not `float_normal_sample`, but I imagine the old state was utilized both for float and double sampling. (I don't think there is necessarily anything wrong with this, but when it comes to BC guarantees, we should call a spade a spade, and say what we are actually preserving, and what we are not. What we seem to have here is "best effort" BC--we won't crash, but the random numbers won't necessarily be the same.)

Reply:
Yes, we would get identical behavior, since I actually haven't utilized the float versions that I'm storing and have used `at::normal_distribution<double>` throughout (since that's what being done in TH right now, THRandom_normal returns a double, and is utilized for both float and double dispatch).